File: frontend/lib/agentResponse.ts
function unescapeJsonString(raw: string): string {
  try {
    return JSON.parse(`"${raw}"`) as string;
  } catch {
    return raw
      .replace(/\\n/g, "\n")
      .replace(/\\t/g, "\t")
      .replace(/\\"/g, '"')
      .replace(/\\\\/g, "\\");
  }
}

function extractQuotedContent(text: string): string | null {
  const m = text.match(/"content"\s*:\s*"/);
  if (!m || m.index === undefined) return null;
  const start = m.index + m[0].length;
  const meta = text.slice(start).match(/"\s*,\s*"metadata"\s*:/);
  if (meta?.index !== undefined) {
    return unescapeJsonString(text.slice(start, start + meta.index));
  }
  return null;
}

/** 从 Agent JSON / Markdown 代码块中取出给用户看的 content */
export function unwrapAgentContent(raw: string): string {
  let text = raw.trim();
  const fence = text.match(/```(?:json)?\s*([\s\S]*?)\s*```/i);
  if (fence?.[1]) {
    text = fence[1].trim();
  }
  const brace = text.indexOf("{");
  if (brace > 0) {
    text = text.slice(brace);
  }

  if (text.startsWith("{") && text.includes('"content"')) {
    try {
      const data = JSON.parse(text) as { content?: unknown };
      if (typeof data.content === "string") {
        return data.content;
      }
    } catch {
      const extracted = extractQuotedContent(text);
      if (extracted) return extracted;
    }
  }

  if (raw.includes('"content"')) {
    const extracted = extractQuotedContent(raw);
    if (extracted) return extracted;
  }

  return raw;
}

File: backend/app/agents/agent_response.py
"""从各 Agent 的 LLM 回复中提取供前端展示的 content 正文。"""
import json
import re


def _normalize_raw(text: str) -> str:
    cleaned = text.strip()
    fence = re.search(r"```(?:json)?\s*([\s\S]*?)\s*```", cleaned, re.IGNORECASE)
    if fence:
        cleaned = fence.group(1).strip()
    brace = cleaned.find("{")
    if brace > 0:
        cleaned = cleaned[brace:]
    return cleaned


def _unescape_json_string(raw: str) -> str:
    try:
        return json.loads(f'"{raw}"')
    except json.JSONDecodeError:
        return (
            raw.replace("\\n", "\n")
            .replace("\\t", "\t")
            .replace('\\"', '"')
            .replace("\\\\", "\\")
        )


def _extract_quoted_content(text: str) -> str | None:
    m = re.search(r'"content"\s*:\s*"', text, re.DOTALL)
    if not m:
        return None
    start = m.end()
    meta = re.search(r'"\s*,\s*"metadata"\s*:', text[start:], re.DOTALL)
    if meta:
        return _unescape_json_string(text[start : start + meta.start()])
    out: list[str] = []
    i = start
    while i < len(text):
        ch = text[i]
        if ch == "\\" and i + 1 < len(text):
            out.append(text[i : i + 2])
            i += 2
            continue
        if ch == '"':
            break
        out.append(ch)
        i += 1
    return _unescape_json_string("".join(out))


def content_from_agent_response(raw: str) -> str:
    """解析 Agent JSON 回复中的 content；兼容 Markdown 代码块与非法 JSON。"""
    if not raw or not raw.strip():
        return raw

    normalized = _normalize_raw(raw)
    try:
        data = json.loads(normalized)
        if isinstance(data, dict) and isinstance(data.get("content"), str):
            return data["content"].strip()
    except json.JSONDecodeError:
        pass

    extracted = _extract_quoted_content(normalized) or _extract_quoted_content(raw)
    if extracted:
        return extracted.strip()

    return raw.strip()
